src/controller/controller.cpp

```diff
@@ -324,67 +273,56 @@ void Controller::onDialogOK(const EidCertificateAndPinInfo& certAndPinInfo)
         onConfirmCommandHandler(certAndPinInfo);
     } else {
         // This should not happen, and when it does, OK should be equivalent to cancel.
-        onPinPadCancel();
+        onDialogCancel();
     }
 }
 
 void Controller::onDialogCancel()
 {
-    REQUIRE_NON_NULL(window)
-
     qDebug() << "User cancelled";
-
-    // Schedule application exit when the UI dialog is destroyed.
-    connect(window, &WebEidUI::destroyed, this, &Controller::exit);
-
     _result = makeErrorObject(RESP_USER_CANCEL, QStringLiteral("User cancelled"));
     writeResponseToStdOut(isInStdinMode, _result, commandType());
-}
-
-void Controller::onPinPadCancel()
-{
-    REQUIRE_NON_NULL(window)
-
-    onDialogCancel();
-    window->quit();
+    exit();
 }
 
 void Controller::onCriticalFailure(const QString& error)
 {
+    emit stopCardEventMonitorThread();
     qCritical() << "Exiting due to command" << std::string(commandType())
                 << "fatal error:" << error;
     _result =
         makeErrorObject(RESP_TECH_ERROR, QStringLiteral("Technical error, see application logs"));
-    writeResponseToStdOut(isInStdinMode, _result, commandType());
     disposeUI();
+    if (qApp->isSafariExtensionContainingApp()) {
+        writeResponseToStdOut(isInStdinMode, _result, commandType());
+    }
     WebEidUI::showFatalError();
+    if (!qApp->isSafariExtensionContainingApp()) {
+        // Write the error response to stdout after showing the fatal error dialog. Chrome will
+        // close the application immediately after this, so the UI dialog may not be visible to the
+        // user.
+        writeResponseToStdOut(isInStdinMode, _result, commandType());
+    }
     exit();
 }
 
 void Controller::exit()
 {
-    if (window) {
-        window->disconnect();
-        window = nullptr;
-    }
+    disposeUI();
     waitForChildThreads();
     emit quit();
 }
 
 void Controller::waitForChildThreads()
 {
-    // Waiting for child threads must not happen in destructor.
-    // See https://tombarta.wordpress.com/2008/07/10/gcc-pure-virtual-method-called/ for details.
-    for (const auto& childThread : childThreads) {
-        auto thread = childThread.second;
-        if (thread) {
-            interruptThread(thread);
-            // Waiting for PIN input on PIN pad may take a long time, call processEvents() so that
-            // the UI doesn't freeze.
-            while (thread->isRunning()) {
-                thread->wait(100); // in milliseconds
-                QCoreApplication::processEvents();
-            }
+    for (auto* thread : findChildren<QThread*>()) {
+        qDebug() << "Interrupting thread" << uintptr_t(thread);
+        thread->disconnect();
+        thread->requestInterruption();
+        ControllerChildThread::waitForControllerNotify.wakeAll();
+        while (thread->isRunning()) {
+            using namespace std::chrono_literals;
+            thread->wait(100ms);
```

Comment:
The 100 millisecond wait made only sense along with event processing, recommend to add it back:

```suggestion
            thread->wait(100ms);
            QCoreApplication::processEvents();
```

Otherwise the while loop does not have a purpose, it is equivalent to just waiting indefinitely.

Reply:
All calls waitForChildThreads() are after disposeUI, no point to process any more events.
and timer makes sure that function does not continue before all thread are ended

Reply:
In this case use plain `thread->wait()` without `while`:

```diff
-        while (thread->isRunning()) {
-            using namespace std::chrono_literals;
-            thread->wait(100ms);
+            thread->wait();
```